caper/nb_subproc_thread.py

```diff
@@ -0,0 +1,182 @@
+import logging
+import time
+from queue import Empty, Queue
+from subprocess import PIPE, STDOUT, CalledProcessError, Popen
+from threading import Thread
+
+logger = logging.getLogger(__name__)
+
+
+class NBSubprocThread(Thread):
+    DEFAULT_POLL_INTERVAL_SEC = 0.01
+
+    def __init__(
+        self,
+        args,
+        cwd=None,
+        stdin=None,
+        on_poll=None,
+        on_stdout=None,
+        on_terminate=None,
+        poll_interval=DEFAULT_POLL_INTERVAL_SEC,
+    ):
+        """Non-blocking STDOUT streaming for subprocess.Popen.
+        Note that STDERR is always redirected to STDOUT.
+
+        This class makes two threads (main and sub).
+        Main thread for subprocess.Popen, sub thread for nb-streaming STDOUT.
+
+        Note that return value of callback functions are used for status or return
+        value (property status, ret) of this thread itself.
+
+        Return value of on_poll and on_std_out are used to update property status.
+        Return value of on_terminate is used to update property ret.
+
+        These are useful to check status of the thread and get the final return
+        value of the function that this thread runs.
+
+        Args:
+            args:
+                List of command line arguments.
+            cwd:
+                subprocess.Popen's cwd.
+            stdin:
+                subprocess.Popen's stdin.
+                Note that subprocess.Popen's stdout/stderr is fixed
+                at subprocess.PIPE/subprocess.STDOUT.
+            on_poll:
+                Callback on every polling.
+                This callback function should take one argument:
+                    - iter (int):
+                        Number of iterations for polling.
+            on_stdout:
+                Callback on every non-empty STDOUT line (ending with backslash n).
+                You can use this to print out STDOUT as well.
+                Return value will replace
+                This callback function should take one argument:
+                    - stdout (string):
+                        New incoming STDOUT line string including backslash n.
+            on_terminate:
+                Callback on terminating a thread.
+                Note that return value of this function will be the final return
+                value (property ret) of this thread after it is done (joined).
+            poll_interval (float):
+                Polling interval in seconds.
+        """
+        super().__init__(
+            target=self._popen,
+            args=(args, cwd, stdin, on_poll, on_stdout, on_terminate),
+        )
+        self._poll_interval = poll_interval
+        self._stdout_list = []
+        self._rc = None
+        self._stop_it = False
+        self._status = None
+        self._ret = None
+
+    @property
+    def stdout(self):
+        return ''.join(self._stdout_list)
+
+    @property
+    def rc(self):
```

Comment:
`rc` and `ret` could be more descriptive.

Reply:
Replaced them with `returncode` and `returnvalue`.